src/python_testing/taglist_and_topology_test_support.py

```diff
@@ -0,0 +1,191 @@
+    return cycles
+
+
+def create_device_type_lists(roots: list[int], endpoint_dict: dict[int, Any]) -> dict[int, dict[int, set[int]]]:
+    """Returns a list of endpoints per device type for each root in the list"""
+    device_types = {}
+    for root in roots:
+        tree_device_types = defaultdict(set)
+        eps = get_all_children(root, endpoint_dict)
+        eps.add(root)
+        for ep in eps:
+            for d in endpoint_dict[ep][Clusters.Descriptor][Clusters.Descriptor.Attributes.DeviceTypeList]:
+                tree_device_types[d.deviceType].add(ep)
+        device_types[root] = tree_device_types
+
+    return device_types
+
+
+def get_direct_children_of_root(endpoint_dict: dict[int, Any]) -> set[int]:
+    root_children = set(endpoint_dict[0][Clusters.Descriptor][Clusters.Descriptor.Attributes.PartsList])
+    direct_children = root_children
+    for ep in root_children:
+        ep_children = set(endpoint_dict[ep][Clusters.Descriptor][Clusters.Descriptor.Attributes.PartsList])
+        direct_children = direct_children - ep_children
+    return direct_children
+
+
+def create_device_type_list_for_root(direct_children, endpoint_dict: dict[int, Any]) -> dict[int, set[int]]:
+    device_types = defaultdict(set)
+    for ep in direct_children:
+        for d in endpoint_dict[ep][Clusters.Descriptor][Clusters.Descriptor.Attributes.DeviceTypeList]:
+            device_types[d.deviceType].add(ep)
+    return device_types
+
+
+def cmp_tag_list(a: Clusters.Descriptor.Structs.SemanticTagStruct, b: Clusters.Descriptor.Structs.SemanticTagStruct):
+    if a.mfgCode != b.mfgCode:
+        return -1 if a.mfgCode < b.mfgCode else 1
+    if a.namespaceID != b.namespaceID:
+        return -1 if a.namespaceID < b.namespaceID else 1
+    if a.tag != b.tag:
+        return -1 if a.tag < b.tag else 1
+    if a.label != b.label:
+        return -1 if a.label < b.label else 1
+    return 0
+
+
+def find_tag_list_problems(roots: list[int], device_types: dict[int, dict[int, set[int]]], endpoint_dict: dict[int, Any]) -> dict[int, TagProblem]:
+    """Checks for non-spec compliant tag lists"""
+    tag_problems = {}
+    for root in roots:
+        for _, endpoints in device_types[root].items():
+            if len(endpoints) < 2:
+                continue
+            for endpoint in endpoints:
+                missing_feature = not bool(endpoint_dict[endpoint][Clusters.Descriptor]
+                                           [Clusters.Descriptor.Attributes.FeatureMap] & Clusters.Descriptor.Bitmaps.Feature.kTagList)
+                if Clusters.Descriptor.Attributes.TagList not in endpoint_dict[endpoint][Clusters.Descriptor] or endpoint_dict[endpoint][Clusters.Descriptor][Clusters.Descriptor.Attributes.TagList] == []:
+                    tag_problems[endpoint] = TagProblem(root=root, missing_attribute=True,
+                                                        missing_feature=missing_feature, duplicates=endpoints)
```

Comment:
Would be great to refactor these into more bool helper variables so that the compound conditions are much more readable.

Reply:
will add to the list, but for another PR - this one is literally just moving code around.